iceoryx_hoofs/include/iceoryx_hoofs/cxx/functional_interface.hpp

```diff
@@ -117,28 +117,32 @@ struct AndThenWithValue
     ///        happens.
     /// @param[in] callable will be called when valid
     /// @return reference to *this
-    Derived& and_then(const and_then_callback_t& callable) & noexcept;
```

Comment:
The alias in line 115 `and_then_callback_t` can be removed. Same for the ones in the struct `OrElseWithValue`.

Reply:
Oh I still need them. The old expected/optional had the requirement to work with unset `cxx::function_ref`, therefore I still have to check this in `and_then`/`or_else`.
And for that I require to convert the callable into the `function_ref` type - if it was already a `function_ref` which was unset I can use the bool operator to check if it is valid and only call it then.

The problem is the functor could be also a lambda which does not have a bool operator and then we cannot perform the check without a compile time error.
To have the same behavior we sadly have to go this way.

But again, another good reason why `cxx::function` & `cxx::function_ref` should not be nullable. It makes the code more error prone and in this case even slower.

Reply:
@mossmaurice I think the convertible check is superfluous since I already check if this thing is invocable as required - the exact same check you perform in the constructor of the function ref.

Furthermore, it is a requirement of the `function_ref` that it is constructible from any callable with the defined signature.

And finally, when the `function_ref` & `function` is no longer nullable we also can get rid of those type aliases and the check inside `and_then` and `or_else`. Then the code become easier and more performant.